### DISCUSSION TITLE
Architecting my application

### DISCUSSION THREAD

**Discussion by erikd** [jtdaugherty/brick#507] [General]
Hi, I am a reasonably experienced Haskell dev but I have never used Brick before.

I have an idea for a Brick app which will display a number of different views of what is basically tablular text data.

Basic needs are as follows:
* Data is read from a file which is updated periodically.
* The view of the data is changed every 15 seconds, in a round robin fashion.
* Keystrokes can cause switches from view to view immediately.

I hoping to get some pointers on how this should be architected. Links to appropriate Haddock docs appreciated.

**Comment by jtdaugherty** [comment]
Hi @erikd - out of curiosity, how familiar are you at this point with brick's design, particularly as it relates to how applications must interact with `brick` itself? For instance, have you read any of [the user guide](https://github.com/jtdaugherty/brick/blob/master/docs/guide.rst) or looked at the demo programs?

**Reply by erikd** [reply]
Reading the user guide now. I did look at https://samtay.github.io/posts/introduction-to-brick (particularly the `snake` example) but that does not compile.

**Reply by jtdaugherty** [reply]
Yeah, that one is a bit old. The demo programs are extensive and you might consider running and looking at those. As you look at the earlier sections of the user guide as well as the demos, I am hoping that some things will become evident w.r.t. how your application needs to be designed, and any introductory comments I could make here probably repeat what is already written down in the guide. If you get to the point where you have specific questions, I would be happy to talk about them here!

**Reply by jtdaugherty** [reply]
Also, the README links to a great many third-party Brick-based programs, some of which are small games like `snake`, so perhaps those could be worth looking at as well.

**Reply by erikd** [reply]
I will delve more deeply into the user guide and the examples over the weekend. Its Friday morning here and work beckons.

**Reply by jtdaugherty** [reply]
Okay, great!

**Reply by erikd** [reply]
Thanks @jtdaugherty . The user guide plus the examples got me sorted out. I have my app rendering a single view of my data and responding to events as it should. All I need to do now is add the new data views.

Basically I needed to:

* Define my state struct.
* Define a `BChan` to supply data to the event handler which can then put it in state struct,
* Define my `appDraw` and `appHandleEvent` functions, Basically did not need the rest.

Thanks again.

**Reply by erikd** [reply]
One little problem I have is that I am using `appChooseCursor = Brick.neverShowCursor` and after I quit the program using `Brick.halt` the terminal that the program was run in also has no cursor. Its there a way to reset the terminal cursor after `Brick.halt` (I am using `customMainWithDefaultVty`)?

I know I can do it by running `reset` in the terminal but that's a bit of a pain.

**Reply by jtdaugherty** [reply]
What platform are you on? That sounds like a bug in the shutdown implementation in Vty. The `neverShowCursor` function is unrelated to shutdown behavior and restoring terminal state; `neverShowCursor` is just a selector for which Brick-rendered cursor gets placed at draw time.

**Reply by erikd** [reply]
i am on Linux.

**Comment by erikd** [comment]
For now, I have fixed the shutdown problem with:
```
  putStrLn ""
  clearScreen
```
With `clearScreen` coming from `System.Console.ANSI`. Unfortunately this only fixes part of the problem. It does not fix the missing cursor.

Also tried `setSGR [ Reset ]` but that did not help.

The only thing that actually worked is `callProcess "reset" []`.

**Reply by jtdaugherty** [reply]
What terminal emulator are you using, and what version of `vty-unix` is installed?

**Comment by erikd** [comment]
Using `mate-terminal` as the emulator and :
```
> jq < dist-newstyle/cache/plan.json | grep vty | grep \"id\"
      "id": "vty-6.2-7fe99ecbb28a8328d7657fa9fdb0bfb415160015f22bdf4fea1be136db944837",
      "id": "vty-crossplatform-0.4.0.0-6ad0c13093c70c759a247514384c83aa14233cb624ab01d1dcea710541ea3eed",
      "id": "vty-unix-0.2.0.0-3dd206aaf8d8d3ed13d5a296247dfe3d827605119333a35963dd1db44ec1b15c",
      "id": "vty-unix-0.2.0.0-e-vty-unix-build-width-table-67138eeb5a8747a6d54a7e68001914dbb0df9331834b7087468bec1a26494900",
```

**Reply by jtdaugherty** [reply]
Would you be willing to try your application on, say, `xterm` or some other terminal emulator without your reset hack and see if the problem is still present?

**Reply by jtdaugherty** [reply]
Also, what is the value of `$TERM`, and what is the output of `infocmp`?

**Reply by jtdaugherty** [reply]
(In `mate-terminal`.)

**Reply by erikd** [reply]
I tired `xterm` and removed the `callProcess "reset" []` and I still get no cursor after `Brick.halt`. Running `reset` on the command line restored the cursor.

In `mate-terminal`:
```
> echo $TERM
xterm-256color
```
which is probably something I set years ago and never changed. `mate-terminal` is a fork of some version of `gnome-terminal` and that now defaults to `xterm-256color`

In `mate-terminal`:
```
> infocmp 
#	Reconstructed via infocmp from file: /usr/share/terminfo/x/xterm-256color
xterm-256color|xterm with 256 colors,
	am, bce, ccc, km, mc5i, mir, msgr, npc, xenl,
	colors#0x100, cols#80, it#8, lines#24, pairs#0x10000,
	acsc=``aaffggiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz{{||}}~~,
	bel=^G, blink=\E[5m, bold=\E[1m, cbt=\E[Z, civis=\E[?25l,
	clear=\E[H\E[2J, cnorm=\E[?12l\E[?25h, cr=\r,
	csr=\E[%i%p1%d;%p2%dr, cub=\E[%p1%dD, cub1=^H,
	cud=\E[%p1%dB, cud1=\n, cuf=\E[%p1%dC, cuf1=\E[C,
	cup=\E[%i%p1%d;%p2%dH, cuu=\E[%p1%dA, cuu1=\E[A,
	cvvis=\E[?12;25h, dch=\E[%p1%dP, dch1=\E[P, dim=\E[2m,
	dl=\E[%p1%dM, dl1=\E[M, ech=\E[%p1%dX, ed=\E[J, el=\E[K,
	el1=\E[1K, flash=\E[?5h$<100/>\E[?5l, home=\E[H,
	hpa=\E[%i%p1%dG, ht=^I, hts=\EH, ich=\E[%p1%d@,
	il=\E[%p1%dL, il1=\E[L, ind=\n, indn=\E[%p1%dS,
	initc=\E]4;%p1%d;rgb:%p2%{255}%*%{1000}%/%2.2X/%p3%{255}%*%{1000}%/%2.2X/%p4%{255}%*%{1000}%/%2.2X\E\\,
	invis=\E[8m, is2=\E[!p\E[?3;4l\E[4l\E>, kDC=\E[3;2~,
	kEND=\E[1;2F, kHOM=\E[1;2H, kIC=\E[2;2~, kLFT=\E[1;2D,
	kNXT=\E[6;2~, kPRV=\E[5;2~, kRIT=\E[1;2C, ka1=\EOw,
	ka3=\EOy, kb2=\EOu, kbeg=\EOE, kbs=^?, kc1=\EOq, kc3=\EOs,
	kcbt=\E[Z, kcub1=\EOD, kcud1=\EOB, kcuf1=\EOC, kcuu1=\EOA,
	kdch1=\E[3~, kend=\EOF, kent=\EOM, kf1=\EOP, kf10=\E[21~,
	kf11=\E[23~, kf12=\E[24~, kf13=\E[1;2P, kf14=\E[1;2Q,
	kf15=\E[1;2R, kf16=\E[1;2S, kf17=\E[15;2~, kf18=\E[17;2~,
	kf19=\E[18;2~, kf2=\EOQ, kf20=\E[19;2~, kf21=\E[20;2~,
	kf22=\E[21;2~, kf23=\E[23;2~, kf24=\E[24;2~,
	kf25=\E[1;5P, kf26=\E[1;5Q, kf27=\E[1;5R, kf28=\E[1;5S,
	kf29=\E[15;5~, kf3=\EOR, kf30=\E[17;5~, kf31=\E[18;5~,
	kf32=\E[19;5~, kf33=\E[20;5~, kf34=\E[21;5~,
	kf35=\E[23;5~, kf36=\E[24;5~, kf37=\E[1;6P, kf38=\E[1;6Q,
	kf39=\E[1;6R, kf4=\EOS, kf40=\E[1;6S, kf41=\E[15;6~,
	kf42=\E[17;6~, kf43=\E[18;6~, kf44=\E[19;6~,
	kf45=\E[20;6~, kf46=\E[21;6~, kf47=\E[23;6~,
	kf48=\E[24;6~, kf49=\E[1;3P, kf5=\E[15~, kf50=\E[1;3Q,
	kf51=\E[1;3R, kf52=\E[1;3S, kf53=\E[15;3~, kf54=\E[17;3~,
	kf55=\E[18;3~, kf56=\E[19;3~, kf57=\E[20;3~,
	kf58=\E[21;3~, kf59=\E[23;3~, kf6=\E[17~, kf60=\E[24;3~,
	kf61=\E[1;4P, kf62=\E[1;4Q, kf63=\E[1;4R, kf7=\E[18~,
	kf8=\E[19~, kf9=\E[20~, khome=\EOH, kich1=\E[2~,
	kind=\E[1;2B, kmous=\E[<, knp=\E[6~, kpp=\E[5~,
	kri=\E[1;2A, mc0=\E[i, mc4=\E[4i, mc5=\E[5i, meml=\El,
	memu=\Em, mgc=\E[?69l, nel=\EE, oc=\E]104\007,
	op=\E[39;49m, rc=\E8, rep=%p1%c\E[%p2%{1}%-%db,
	rev=\E[7m, ri=\EM, rin=\E[%p1%dT, ritm=\E[23m, rmacs=\E(B,
	rmam=\E[?7l, rmcup=\E[?1049l\E[23;0;0t, rmir=\E[4l,
	rmkx=\E[?1l\E>, rmm=\E[?1034l, rmso=\E[27m, rmul=\E[24m,
	rs1=\Ec\E]104\007, rs2=\E[!p\E[?3;4l\E[4l\E>, sc=\E7,
	setab=\E[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e48;5;%p1%d%;m,
	setaf=\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e38;5;%p1%d%;m,
	sgr=%?%p9%t\E(0%e\E(B%;\E[0%?%p6%t;1%;%?%p5%t;2%;%?%p2%t;4%;%?%p1%p3%|%t;7%;%?%p4%t;5%;%?%p7%t;8%;m,
	sgr0=\E(B\E[m, sitm=\E[3m, smacs=\E(0, smam=\E[?7h,
	smcup=\E[?1049h\E[22;0;0t, smglp=\E[?69h\E[%i%p1%ds,
	smglr=\E[?69h\E[%i%p1%d;%p2%ds,
	smgrp=\E[?69h\E[%i;%p1%ds, smir=\E[4h, smkx=\E[?1h\E=,
	smm=\E[?1034h, smso=\E[7m, smul=\E[4m, tbc=\E[3g,
	u6=\E[%i%d;%dR, u7=\E[6n, u8=\E[?%[;0123456789]c,
	u9=\E[c, vpa=\E[%i%p1%dd,
```

**Reply by jtdaugherty** [reply]
> I tired xterm and removed the callProcess "reset" [] and I still get no cursor after Brick.halt.

Yikes, I find this *really* surprising since `xterm` has been supported in Vty for years and reset behavior has not been broken at any point that I can remember. Huh. Are you by chance doing anything with Vty yourself in your application, like making  your own Vty handle?

**Reply by erikd** [reply]
Nope, just running `customMainWithDefaultVty` but I have done:
```
 App
   { ..
    , appAttrMap = const $ Brick.attrMap Vty.defAttr []
   }
```
Is there something I should be doing instead.

**Reply by jtdaugherty** [reply]
There is nothing else you should be doing; the expected behavior is that Vty (which Brick is built on) will restore the terminal state properly on exit, including the display of the cursor. That this isn't happening properly is highly unexpected but I don't think it's caused by anything you're doing.

I'm not sure what to do to investigate further, though. The fact that it's happening on *both* `xterm` and `mate-terminal` suggests that it isn't specific to one of those. I figured it might be possible that the reset sequences that Vty is emitting don't work for `mate-terminal`, but it would seem that on your system they also don't work for `xterm`, which I find extremely surprising.

**Reply by kquick** [reply]
Do the demo programs that come with brick have the same behavior?

**Reply by jtdaugherty** [reply]
Thanks, @kquick - got any ideas about what might be going on here?

**Reply by kquick** [reply]
It gives us another set of datapoints.  

I've tried a couple of my Brick apps in mate-terminal and don't see this issue.

If somehow the termcaps got written with some unusual alternative then it might cause unusual behavior.  

@erikd, what linux distro and brick version?

**Reply by jtdaugherty** [reply]
@kquick for what it's worth: Vty's `shutdown` calls `releaseDisplay` and then `releaseTerminal`, and for XTerm-based terminals, those trigger the emission of the `rmcup` and `cnorm` capability strings (for `releaseDisplay`) and `sgr0` and `cnorm` for `releaseTerminal`. So if any of those are off, that could explain this.

**Reply by erikd** [reply]
Wow it gets weird.

> Do the demo programs that come with brick have the same behavior?

Yes, when run on the machine I was working on. On a second machine correctly restores the cursor, both with one of the demo programs and also with my program.

This looks like some issue with my machine.

> what linux distro and brick version?

Debian Testing on both machines. I have run `debums -s` on the machine with the problems, but nothing was reported.

**Reply by jtdaugherty** [reply]
Is there perhaps something different about your shell configuration across those machines? Something present in your `.bashrc` (say) that might be contributing to this?

**Reply by erikd** [reply]
Same `.bashrc`s on both machines.

**Comment by erikd** [comment]
Thanks @jtdaugherty and @kquick . I think I will put this cursor issue down to something being astray on my particular machine,

**Comment by erikd** [comment]
My basic answer is https://github.com/jtdaugherty/brick/discussions/507#discussioncomment-9318778 .

Closing this.
